cmd/goal/interact.go

```diff
@@ -0,0 +1,546 @@
+	for k := range sch {
+		names = append(names, k)
+	}
+	return
+}
+
+func (sch appInteractSchema) ToStateSchema() (schema basics.StateSchema) {
+	for name, entry := range sch {
+		size := uint64(1)
+		if entry.Map.Kind != "" {
+			if entry.Size < 0 {
+				reportErrorf("entry %s size %d < 0", name, entry.Size)
+			}
+			size = uint64(entry.Size)
+		}
+		switch entry.Kind {
+		case "int", "integer":
+			schema.NumUint += size
+		default:
+			schema.NumByteSlice += size
+		}
+	}
+	return
+}
+
+type appInteractState struct {
+	Global appInteractSchema `json:"global"`
+	Local  appInteractSchema `json:"local"`
+}
+
+type appInteractHeader struct {
+	// map procedure name -> procedure
+	Execute map[string]appInteractProc `json:"execute"`
+
+	Query appInteractState `json:"query"`
+}
+
+func (hdr appInteractHeader) ProcList() string {
+	help := make(map[string]appInteractDatum)
+	for k, v := range hdr.Execute {
+		help[k] = v
+	}
+	return helpList(help)
+}
+
+func (hdr appInteractHeader) ProcNames() (names []string) {
+	for k := range hdr.Execute {
+		names = append(names, k)
+	}
+	return
+}
+
+func parseAppHeader() (header appInteractHeader) {
+	if appHdr == "" {
+		reportErrorf("No header file provided")
+	}
+
+	f, err := os.Open(appHdr)
+	if err != nil {
+		reportErrorf("Could not open app header file %s: %v", appHdr, err)
+	}
+
+	dec := json.NewDecoder(f)
+	err = dec.Decode(&header)
+	if err != nil {
+		reportErrorf("Could not decode app header JSON file %s: %v", appHdr, err)
+	}
+
+	return
+}
+
+// TODO print help correctly when --help is passed but procedure/state name is given
+// TODO complain when unknown flags are given
+
+var appExecuteCmd = &cobra.Command{
+	Use:   "execute",
+	Short: "Execute a procedure on an application",
+	Args:  cobra.MinimumNArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		dataDir := ensureSingleDataDir()
+		client := ensureFullClient(dataDir)
+
+		header := parseAppHeader()
+		proc, ok := header.Execute[args[0]]
+		if !ok {
+			reportErrorf("Unknown procedure name %s.\nDefined procedures in %s:\n%s", args[0], appHdr, header.ProcList())
+		}
+
+		if proc.Create != (appIdx == 0) {
+			reportErrorf("Procedure %s has the create flag set to %v, but application ID set to %d", args[0], proc.Create, appIdx)
+		}
+
+		procFlags := cmd.Flags()
+		procFlags.SetInterspersed(true)
+		procArgs := make(map[string]interface{})
+		procAccounts := make(map[string]*string)
+		procApps := make(map[string]*uint64)
+		for _, arg := range proc.Args {
+			switch arg.Kind {
+			case "int", "integer":
+				procArgs[arg.Name] = procFlags.Uint64(arg.Name, 0, arg.Help)
+			default:
+				procArgs[arg.Name] = procFlags.String(arg.Name, "", arg.Help)
+			}
+		}
+		for _, account := range proc.Accounts {
+			procAccounts[account.Name] = procFlags.String(account.Name, "", account.Help)
+		}
+		for _, app := range proc.ForeignApps {
+			procApps[app.Name] = procFlags.Uint64(app.Name, 0, app.Help)
```

Comment:
It would be worth peeking at the implementation of the type returned by `cmd.Flags()` to make sure that it's safe to pass arbitrary input to these functions. Like, it would be bad if opening an interface could execute arbitrary code through flag parsing (unlikely, but worth a quick check in `cobra`).

Reply:
Took a look at `pflag`, which is what `cobra` uses--couldn't find anything too dangerous (it seems like the untrusted values are the name and help, which are just used to key things during `Parse`). While we're on this subject, though, do we need to have a plan against adversarial JSON spec files? i.e., do we need to handle a case where someone sticks a terminal control character in the help text, or puts null bytes or spaces in either?

Reply:
I do think we need to handle that case, yes

Reply:
From discussion: will be heavily restricting what strings are allowed to prevent abuse. (Later commits can broaden this set.)